horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/RunServiceImpl.java

```diff
@@ -899,22 +906,15 @@ private void onCalculateDataSets(String param) {
       String token = parts[1];
       String role = parts[2];
 
+      Run run = null;
       try (@SuppressWarnings("unused") CloseMe h1 = roleManager.withRoles(em, role);
            @SuppressWarnings("unused") CloseMe h2 = roleManager.withToken(em, token)){
-         Run run = Run.findById(runId);
-         if (run != null) {
-            DataSet dataSet = new DataSet();
-            dataSet.start = run.start;
-            dataSet.stop = run.stop;
-            dataSet.description = run.description;
-            dataSet.testid = run.testid;
-            dataSet.run = run;
-            dataSet.data = run.data;
-            dataSet.owner = run.owner;
-            dataSet.access = run.access;
-            dataSet.persist();
-            Util.publishLater(tm, eventBus, DataSet.EVENT_NEW, dataSet);
-         }
+          run = Run.findById(runId);
+      } catch (Exception e) {
+         log.errorf("Calculate DataSet issue: exception %s", e);
+      }
+      if (run != null) {
+         transform(run, role, token);
```

Comment:
Looks like you're entering authenticated block once above (to find the run) and then second time inside `transform`. In addition to that the code is running anonymous when fetching run schemas - that most likely won't work for private runs.

Reply:
thanks for spotting that.

Reply:
Actually I realized that the transformation, as well as `onCalculateTags` shouldn't be executed with uploading user permissions. So far we've been running everything with Public access so it worked anyway, but we should not have to expose (private) Transformers or other entities to the uploader role. We need to see everything related and the simple solution is to use the `horreum.system` role for that. I already did that for Labels materialization (tunneling privileges would be even more complicated in there).

The result of transformation should be deterministic, so an attacker who would trigger recalculation by executing `pg_notify()` in SQL should not cause any harm.

I am not asking you to change this here and now; just keep it in mind when rebasing on changes I will introduce in the mainline.

Reply:
is `roleManager.withRoles(em, "horreum.system")` supposed to work ? I am seeing row level security policy errors.

Reply:
No, the second argument here is `signedRoles` - concatenation of cryptographically signed ones. You want to use roleManager.withRoles(em, Collections.singleton(Roles.HORREUM_SYSTEM))`, or preferably make the method non-private and annotate it with `@WithRoles(extras = Roles.HORREUM_SYSTEM)`.